Feature/request v3

sduchesneau opened the pull request:
This allows sending the full spkg instead of just the modules in the substreams request.
The server can do the mangling of replacing the network and params with a provided value.

sduchesneau committed:
add support for sf.substreams.rpc.v3

sduchesneau committed:
fix naming of protobuf generated modules

sduchesneau committed:
pass down params and network overrides when sending v2 requests

sduchesneau committed:
fix test case flag matching

maoueh commented on app/tier1.go:
```diff
@@ -21,8 +21,7 @@ import (
 	"github.com/streamingfast/substreams/metrics"
-	"github.com/streamingfast/substreams/pb/sf/substreams/rpc/v2/pbsubstreamsrpcconnect"
-	ssconnect "github.com/streamingfast/substreams/pb/sf/substreams/rpc/v2/pbsubstreamsrpcconnect"
+	ssconnect "github.com/streamingfast/substreams/pb/sf/substreams/rpc/v2/pbsubstreamsrpcv2connect"
```
Even if it's long, I think we should still use `pbsubstreamsrpcv2connect`, IMO.

maoueh commented on client/client.go:
```diff
@@ -88,7 +94,7 @@ func (c *SubstreamsClientConfig) MarshalLogObject(encoder zapcore.ObjectEncoder)
 type InternalClientFactory = func() (cli pbssinternal.SubstreamsClient, closeFunc func() error, callOpts []grpc.CallOption, headers Headers, err error)
 
-func NewSubstreamsClientConfig(endpoint string, authToken string, authType AuthType, insecure bool, plaintext bool, agent string) *SubstreamsClientConfig {
+func NewSubstreamsClientConfig(endpoint string, authToken string, authType AuthType, insecure bool, plaintext bool, agent string, forceV2 bool) *SubstreamsClientConfig {
```
We should probably think about using a struct as this function parameters, we added some parameters over time and I think we probably will add more after this one, up to you to refactor or not.

This will now be a breaking change also for people consuming `substreams` as a `substreams-sink` library.

maoueh replied:
Should we specify forced version instead of v3, like `forceVersion string` (or enum would be better).

maoueh commented on client/client.go:
```diff
@@ -278,8 +287,18 @@ func NewSubstreamsClient(config *SubstreamsClientConfig) (cli pbsubstreamsrpc.St
+	return
+}
+
+func NewSubstreamsClients(config *SubstreamsClientConfig) (cliV2 pbsubstreamsrpcv2.StreamClient, cliV3 pbsubstreamsrpcv3.StreamClient, closeFunc func() error, callOpts []grpc.CallOption, headers Headers, err error) {
```
Should it be `NewSubstreamsClientConn(...) ...` and caller would be responsible of :

```
cliV2 := pbsubstreamsrpcv2.NewStreamClient(conn)
cliV3 := pbsubstreamsrpcv3.NewStreamClient(conn)
```

maoueh commented on docs/release-notes/change-log.md:
```diff
@@ -23,6 +23,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 * Added support for foundational-store (in wasmtime and wazero).
+* Added support for new 'sf.substreams.rpc.v3.Stream/Blocks' endpoint that sends the full '.spkg' data with params and network value, so the client does not need to do any mangling.
+  * This requires the substreams server to support it (under `/sf.substreams.rpc.v2.Stream/*` location).
```
Hmmm `...rpc.v3`?

maoueh commented on docs/release-notes/change-log.md:
```diff
@@ -23,6 +23,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 * Added support for foundational-store (in wasmtime and wazero).
+* Added support for new 'sf.substreams.rpc.v3.Stream/Blocks' endpoint that sends the full '.spkg' data with params and network value, so the client does not need to do any mangling.
+  * This requires the substreams server to support it (under `/sf.substreams.rpc.v2.Stream/*` location).
+  * On the `run`, `gui` and `sink` commands, the `--force-v2` flag is available to keep the previous protocol; the `v2` endpoint will also be tried as fallback if the server responds with 404 or MethodNotAllowed.
```
To be thought of in relation to maybe yet a future v4.

maoueh commented on pb/sf/substreams/rpc/v2/substreams.go:
```diff
@@ -1,4 +1,4 @@
-package pbsubstreamsrpc
+package pbsubstreamsrpcv2
```
We should think about this rename if truly necessary, we could change only our internal usage, that would make the Protobuf fully the same as before.

maoueh commented on pb/sf/substreams/rpc/v3/request.go:
```diff
@@ -0,0 +1,92 @@
+		ProductionMode:                      r.ProductionMode,
+		OutputModule:                        r.OutputModule,
+		Modules:                             r.Package.Modules,
+		DebugInitialStoreSnapshotForModules: r.DebugInitialStoreSnapshotForModules,
```
We should probably take the time to remove this one entierly from v3, it think it has ben unsupported for a long time now.

maoueh commented on pb/sf/substreams/rpc/v3/request.go:
```diff
@@ -0,0 +1,92 @@
+	}
+	out := make(map[string]string)
+	for _, p := range in {
+		kv := strings.SplitN(p, "=", 2)
+		if len(kv) != 2 {
+			return nil, fmt.Errorf("invalid parameter format: %s", p)
+		}
```
There is `strings.Cut` now that exists to make that "clearer", in this particular case, it would be mostly equivalent.

maoueh commented on proto/sf/substreams/rpc/v3/service.proto:
```diff
@@ -0,0 +1,72 @@
+
+option go_package = "github.com/streamingfast/substreams/pb/sf/substreams/rpc/v3;pbsubstreamsrpcv3";
+
+service Stream {
+  rpc Blocks(Request) returns (stream sf.substreams.rpc.v2.Response); // v3 request still sends v2 response
+}
+
+message Request {
+  int64 start_block_num = 1;
+  string start_cursor = 2;
+  uint64 stop_block_num = 3;
```
Let's ensure we properly document all fields properly with the right amount of details, like precedence of `start_cursor` if both `start_block_num` and `start_cursor` are provided as well as how to follow chain's head.

maoueh commented on proto/sf/substreams/rpc/v3/service.proto:
```diff
@@ -0,0 +1,72 @@
+import "sf/firehose/v2/firehose.proto";
+import "sf/substreams/rpc/v2/service.proto";
+import "sf/substreams/v1/clock.proto";
+import "sf/substreams/v1/modules.proto";
+import "sf/substreams/v1/package.proto";
+
+option go_package = "github.com/streamingfast/substreams/pb/sf/substreams/rpc/v3;pbsubstreamsrpcv3";
+
+service Stream {
+  rpc Blocks(Request) returns (stream sf.substreams.rpc.v2.Response); // v3 request still sends v2 response
+}
+
+message Request {
```
Documentation

maoueh commented on proto/sf/substreams/rpc/v3/service.proto:
```diff
@@ -0,0 +1,72 @@
+  repeated string dev_output_modules = 13;
+
+  // Progress_messages_interval_ms is the interval between progress messages, in milliseconds (minimum: 500, default: start at 500ms and ramp up to 5000ms within 1min)
+  uint64 progress_messages_interval_ms = 14;
```
Should we take the occassion to change the default?

maoueh commented on tools/prometheus-exporter.go:
```diff
@@ -50,6 +51,7 @@ func init() {
 	prometheusCmd.Flags().String("substreams-api-key-envvar", "SUBSTREAMS_API_KEY", "Name of variable containing Substreams Api Key")
 	prometheusCmd.Flags().BoolP("insecure", "k", false, "Skip certificate validation on GRPC connection")
 	prometheusCmd.Flags().BoolP("plaintext", "p", false, "Establish GRPC connection in plaintext")
+	prometheusCmd.Flags().Bool("force-v2", false, "Force using V2 API")
```
This is missing details, unclear what this V2 API is about.

Side question, should we move `prometheus` to `substreams sink ...`? Maybe it's too specific to our needs though.

codegen-sh[bot] committed:
Apply PR review comments: update import aliases and client connection pattern

- Replace ssconnect with pbsubstreamsrpcv2connect import alias
- Replace ssconnectv3 with pbsubstreamsrpcv3connect import alias
- Refactor from NewSubstreamsClients to NewSubstreamsClientConn pattern
- Update client usage across sink, tools, and tui2 packages
- Fix changelog documentation reference from v2 to v3 endpoint

codegen-sh[bot] committed:
Fix failing check suites: update Go version to 1.24.x in workflows

- Update GitHub Actions workflow to use Go 1.24.x to match go.mod requirements
- Fix import alias formatting in app/tier1.go for consistency

This addresses the Go version mismatch that was causing compilation failures
in the Build and Test workflow.

Co-authored-by: Stéphane Duchesneau <stephane.duchesneau@gmail.com>

codegen-sh[bot] committed:
Fix compilation issues and update GitHub Actions

- Remove unused imports in client/client.go (pbsubstreamsrpcv2 and pbsubstreamsrpcv3)
- Update GitHub Actions to use more recent versions:
  - actions/setup-go@v2 -> actions/setup-go@v5 (to support Go 1.24.x)
  - actions/checkout@v3 -> actions/checkout@v4
  - actions/cache@v3 -> actions/cache@v4

These changes fix the compilation errors and ensure the CI workflows
can properly handle the Go 1.24.x version requirement.

Co-authored-by: Stéphane Duchesneau <stephane.duchesneau@gmail.com>

codegen-sh[bot] committed:
Fix Buf breaking change check by excluding go_package changes

The v3 protocol feature intentionally changes the go_package option from
'pbsubstreamsrpc' to 'pbsubstreamsrpcv2' to distinguish the v2 protocol
from the new v3 protocol. This is a legitimate architectural change that
should be allowed.

Co-authored-by: Stéphane Duchesneau <stephane.duchesneau@gmail.com>

sduchesneau committed:
fix all requested changes in PR: code improvements, flag
